### PR TITLE
feat: lazy retrievals

### PR DESCRIPTION
**Pull request by tchardin**
This is a rather significant amount of changes as progressively retrieving keys required a bit of refactoring.

The aim of this PR is to give the ability to call `Load(<root>)` for retrieving the index of all entries of a given DAG root from a provider and loading the necessary funds to later retrieve parts or all of the DAG. Therefore, the Load RPC allows the following:
- `Load(<root>)` --> Retrieve the index (map with all the keys and entries without resolving the links) and load funds to pay for retrieving all of them with a select provider.
- `Load(<root>/*)` --> Retrieve the index with all linked data
- `Load(<root>/<key>)` --> Retrieve the index and only the linked data for the given key

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -81,3 +81,5 @@ require (
 )
 
 replace github.com/filecoin-project/filecoin-ffi => ./extern/filecoin-ffi
+
+replace github.com/filecoin-project/go-data-transfer => ../go-data-transfer
```
Comment (gallexis) [inline go.mod:85]: @tchardin should we keep that?

Comment (tchardin) [reply go.mod:85]: Good catch. Will remove it, thanks!

**Review by gallexis — APPROVED** [review]
It doesn't work anymore with `blockGen.Next()` or it's just better to use `CreateRandomBlock()`?

**Comment by tchardin** [comment]
> It doesn't work anymore with `blockGen.Next()` or it's just better to use `CreateRandomBlock()`?

It works but was logging a ton of errors so I just changed it so it would be cleaner

**Pull request merged**